### PR TITLE
RUM-16113: Prevent inactive views from overwriting last_view_event

### PR DESCRIPTION
**Pull request by ambushwork**
### What does this PR do?

Prevents view events of views that are no longer displayed from overwriting `last_view_event`.                                                                                                                                                                                                                 
`last_view_event` is the snapshot crash handlers (NDK, ANR, late crash reporting) read to attribute a crash to a RUM view.                            
`RumDataWriter.onDataWritten` used to overwrite it on **every** `ViewEvent`, so an event emitted by an already-replaced view became the crash  attribution target.                                                                                                                                     
                                                                                                                                                         
`RumDataWriter` now resolves the current view from the RUM feature context and persists the event only if it belongs to that view, or to a view that started late.

### Motivation

RUM-16113


### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](../CONTRIBUTING.md) doc)

### CODE REVIEW

**Comment by ambushwork** [comment]
@codex review

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt [hunk 1]
```diff
@@ -83,7 +83,19 @@ internal class RumDataWriter(
...
+    }
+
+    @WorkerThread
+    private fun onViewEventWritten(data: ViewEvent, rawData: ByteArray) {
+        val rumContext = sdkCore.getFeatureContext(Feature.RUM_FEATURE_NAME, useContextThread = false)
+        val currentViewId = rumContext[RumContext.VIEW_ID] as? String
+        val currentViewStartedAt = rumContext[RumContext.VIEW_TIMESTAMP] as? Long ?: 0L
+        val isCurrentView = data.view.id == currentViewId
+        val isNewerView = data.date > currentViewStartedAt
+        if (isCurrentView || isNewerView) {
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt]: fixed

[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt]: fixed

**Comment by ambushwork** [comment]
@codex review

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt [hunk 2]
```diff
@@ -83,7 +87,22 @@ internal class RumDataWriter(
...
+    }
+
+    @WorkerThread
+    private fun onViewEventWritten(data: ViewEvent, rawData: ByteArray) {
+        val viewId = data.view.id
+        val isPersistedView = viewId == lastPersistedViewId
+        val isNewView = seenViewIds.add(viewId)
+        while (seenViewIds.size > MAX_TRACKED_VIEW_IDS) {
+            val oldest = seenViewIds.firstOrNull() ?: break
+            seenViewIds.remove(oldest)
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt]: fixed

**Comment by ambushwork** [comment]
@codex review

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt [hunk 3]
```diff
@@ -83,7 +85,17 @@ internal class RumDataWriter(
...
         when (data) {
-            is ViewEvent -> sdkCore.writeLastViewEvent(rawData)
+            is ViewEvent -> onViewEventWritten(data, rawData)
+        }
+    }
+
+    @WorkerThread
+    private fun onViewEventWritten(data: ViewEvent, rawData: ByteArray) {
+        val isPersistedView = data.view.id == lastPersistedViewId
+        val isViewStart = data.dd.documentVersion == FIRST_VIEW_DOCUMENT_VERSION
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt]: fixed

**Comment by ambushwork** [comment]
@codex review

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt [hunk 4]
```diff
@@ -36,11 +36,17 @@ internal class RumDataWriter(
...
+    private var currentViewId: String? = null
+
     // region DataWriter
 
     @WorkerThread
     @Suppress("ReturnCount")
     override fun write(writer: EventBatchWriter, element: Any, eventType: EventType): Boolean {
+        if (element is ViewEvent) {
+            onViewEventSubmitted(element)
+        }
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt:48]: it needs the storage thread to be saturated at the exact moment a view starts, plus a native crash during that view. Realistically  that's a low-single-digit-per-million-sessions event, and only for apps that already saturate the SDK's storage queue.

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt [hunk 5]
```diff
@@ -83,7 +89,23 @@ internal class RumDataWriter(
...
+        synchronized(this) {
+            if (data.dd.documentVersion == FIRST_VIEW_DOCUMENT_VERSION) {
+                currentViewId = data.view.id
+            }
+        }
+    }
+
+    @WorkerThread
+    private fun onViewEventWritten(data: ViewEvent, rawData: ByteArray) {
+        if (data.view.id == currentViewId) {
```
Comment (satween) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt:107]: shouldn't it be also syncronized?

Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt:107]: this function is already called inside a `syncronized` block in `write` function, so no race condition can happen here.

[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt:107]: this is false, codex confused a scope key with the event view id

**Review by satween — APPROVED** [review]

**Review by jonathanmos — APPROVED** [review]

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt [hunk 6]
```diff
@@ -92,6 +114,8 @@ internal class RumDataWriter(
     companion object {
         val EMPTY_BYTE_ARRAY = ByteArray(0)
 
+        internal const val FIRST_VIEW_DOCUMENT_VERSION = 2L
```
Comment (0xnm) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt:117]: This part is a bit fragile and relies on the fact that we increase document version _before_ we create the event.

**Review by 0xnm — APPROVED** [review]
It seems the following PR description is wrong:

> RumDataWriter now resolves the current view from the RUM feature context ...

Otherwise lgtm.

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt (hunk 6 above)
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/domain/RumDataWriter.kt:117]: it's true, but comparing to other solutions this one has the smallest blast radius and fewest edge cases. To make it less  fragile we can define the initial document version in `RumViewScope` instead of magic number 1, then refer it here. Also add unit test to cover the fact that we rely on the increasing of the doc version before the event creation. WDYH
